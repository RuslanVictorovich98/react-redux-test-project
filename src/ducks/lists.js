import { call, put, takeLatest } from 'redux-saga/effects';

// ActionsType
export const EDIT_LIST = 'EDIT_LIST';
export const EDIT_CATEGORY = 'EDIT_CATEGORY';
export const CREATE_LIST = 'CREATE_LIST';
export const FIND_TO_NAME = 'FIND_TO_NAME';
export const SAVE_FIND_DATA = 'SAVE_FIND_DATA';
export const FILTER_LIST = 'FIND_TO_NAME';
export const GETTING_PRODUCTS = 'GETTING_PRODUCTS' ;
export const FETCH_USERS_REQUEST = 'FETCH_USERS_REQUEST';
export const FIRTS_RENDER_PUTHNAME = 'FIRTS_RENDER_PUTHNAME';
export const FIND_TO_NAME_PATHNAME = 'FIND_TO_NAME_PATHNAME';

// ActionsCreator
export const create = (event) => {
    return {
        type: CREATE_LIST,
        payload: event
    }
}

export const category = (event) => {
    return {
        type: EDIT_CATEGORY,
        payload: event,
    }
}

export const saveFindToName = (event) => {
    return {
        type: SAVE_FIND_DATA,
        payload: event
    }
}

export const findToName = (event) => {
    return {
        type: FIND_TO_NAME,
        payload: event,
    }
}

export const gettingProducts = (event) => {
    return {
        type: GETTING_PRODUCTS,
        payload: event,
    }
}

export const fetchUsersRequest = () => {
    return {type: FETCH_USERS_REQUEST}
}

export const firstRenderPuthname = (event) => {
    return {
        type: FIRTS_RENDER_PUTHNAME,
        payload: event
    }
}

export const findToNamePathname = (event) => {
    return {
        type: FIND_TO_NAME_PATHNAME,
        payload: event
    }
}

// Reducer 
const list =  {
    products: [],
    productsFilter: [],
    productsFind: [],
    productsPuthname: [],
    productsPuthnameFind: [],
    findData: ''
};

export function saveFindData (state = list.findData, action) {
    switch (action.type) {
        case SAVE_FIND_DATA:
            // console.log(action.payload);

            return action.payload;

        default: 
            return state;
    }
}

export function mainList (state = list.products, action) { // main
    switch (action.type) {
        case CREATE_LIST:
            list.productsFilter = action.payload;
            return action.payload;

        default:
            return state;
    }
}

export function listReducer (state = list.productsFilter, action) { // list data
    switch (action.type){

        case CREATE_LIST:
            list.productsFilter = action.payload;
            return action.payload;

        case GETTING_PRODUCTS: 
            return action.payload;
            

        case EDIT_CATEGORY:
            return list.productsFilter.filter(num => num.bsr_category.toLowerCase() === action.payload.toLowerCase());

        default:
            return state;
    }
}

export function liftFilter  (state = list.productsFind, action) { // list data after filter
    switch (action.type) {

        case FIND_TO_NAME: 
            return action.payload;

        default:
            return state;
    }
}

export function firstProductsPuthname (state = list.productsPuthname, action) { // data at first load page by category
    switch (action.type) {
        
        case FIRTS_RENDER_PUTHNAME:
            console.log(action.payload);
            
            return list.productsFilter.filter(num => num.bsr_category.toLowerCase() === action.payload.toLowerCase());

        default:
            return state;
    }
}

export function firstProductsPuthnameFind (state = list.productsPuthnameFind, action) { // data at first load page by find
    switch(action.type) {

        case FIND_TO_NAME_PATHNAME: 
            return action.payload;

        default: 
            return state;
    }
    
}

// Saga
export function* fetchUser() {
  const url = 'http://localhost:3000/data.json';
  try {
    const data = yield call(() => {
              return fetch(url)
                      .then(res => res.json())
              })
    yield put (create(data.products));
  } catch (error) {
    console.error('fetchUsers return erros: ' + error);
  }        
 }
 
 export default function* GetProduct() {
    yield takeLatest(FETCH_USERS_REQUEST, fetchUser);
  }

