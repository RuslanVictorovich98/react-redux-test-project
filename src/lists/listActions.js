export const EDIT_LIST = 'EDIT_LIST';
export const EDIT_CATEGORY = 'EDIT_CATEGORY';
export const CREATE_LIST = 'CREATE_LIST';
export const FIND_TO_NAME = 'FIND_TO_NAME';
export const FILTER_LIST = 'FIND_TO_NAME';
export const GETTING_PRODUCTS = 'GETTING_PRODUCTS' ;
export const FETCH_USERS_REQUEST = 'FETCH_USERS_REQUEST';
export const FIRTS_RENDER_PUTHNAME = 'FIRTS_RENDER_PUTHNAME';
export const FIND_TO_NAME_PATHNAME = 'FIND_TO_NAME_PATHNAME';






