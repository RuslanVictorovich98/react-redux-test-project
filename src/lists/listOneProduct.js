import React from 'react';
import { connect } from 'react-redux';

class ListOneProduct extends React.Component {

    returnProduct = () => {
        const product = this.props.product;
        return (
            <div className="content">
            <img src={product.img}/><br/>
            <span ><strong>Имя: </strong>{product.name} </span> <br/>
            <span ><strong>Бренд: </strong>{product.brand} </span> <br/>
            <span ><strong>Категория: </strong>{product.bsr_category} </span> <br/>
            <span ><strong>Цена: </strong>{product.price} </span> <br/>
            <span ><strong>Звезд: </strong>{product.stars} </span> <br/>
            </div>
        );
    }

    // sreturnProduct = () => {
    //     const product = this.props.product;
    //     return this.props.main.map((product) => {
    //     return (
    //         <div className="content">
    //         <img src={product.img}/><br/>
    //         <span ><strong>Имя: </strong>{product.name} </span> <br/>
    //         <span ><strong>Бренд: </strong>{product.brand} </span> <br/>
    //         <span ><strong>Категория: </strong>{product.bsr_category} </span> <br/>
    //         <span ><strong>Цена: </strong>{product.price} </span> <br/>
    //         <span ><strong>Звезд: </strong>{product.stars} </span> <br/>
    //         </div>
    //     )
    //     });
    // }



    render() {
        return(
            <div>
                {this.returnProduct()}
            </div>

        )
    }
}

const mapStateToProps = (state) => {
    return {
        list: state.list,
        main: state.main,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
    }
}

export default connect (mapStateToProps, mapDispatchToProps) (ListOneProduct);