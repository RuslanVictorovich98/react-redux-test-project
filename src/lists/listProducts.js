import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ListOneProduct from './listOneProduct';

import {createBrowserHistory} from 'history';
const history = createBrowserHistory();

class ListProducts extends React.Component {
    returnProducts = () => {
        const {firstPathname} = this.props;

        if (firstPathname === 'Home & Kitchen' || firstPathname === 'Sports & Outdoors' || firstPathname === 'Health & Personal Care' || firstPathname === 'Baby Products') {
            if (this.props.find === false) {
                return this.props.firstProductsPuthname.map((product, i) => {
                    return(<ListOneProduct key={i} product={product} />)
                })
            } else {
                return this.props.firstProductsPuthnameFind.map((product, i) => {
                    return <ListOneProduct key={i} product={product} />
                })
            }
        } else if (firstPathname === 'All category' ) {

            if (this.props.find === false) { 
                return this.props.list.map((product, i) => {
                    return (<ListOneProduct key={i} product={product} />)
                })
            } else {
                return this.props.listFilter.map((product, i) => {
                    return <ListOneProduct key={i} product={product} />
                })
            }
        } else return <div className="content-loading">loading...</div>
    }   

    componentDidMount () {
        this.setState({firstHistory: history.location.search.slice(1)}); // save search data in find
    }

    render() {
        return(
            <div className="App-content">
              {this.returnProducts()}  
            </div>
        )
    }
}

ListProducts.propTypes = {
    list: PropTypes.array,
    listFilter: PropTypes.array,
    find: PropTypes.bool
}

const mapStateToProps = (state) => {
    return {
        list: state.list,
        listFilter: state.listFilter,
        firstProductsPuthname: state.firstProductsPuthname,
        firstProductsPuthnameFind: state.firstProductsPuthnameFind,
    }
}

export default connect (mapStateToProps) (ListProducts);

/*
returnFirstLoadProductsFind = () => {

        if (this.props.firstPathname === false) {
            return <div className="content-loading">loading...</div>
        } else if (this.props.firstPathname === 'No category') { 

            let filterMin = this.props.list.filter(num => {
                return num.name.toLowerCase().indexOf(this.state.firstHistory) >= 0 ? num : null;
            })
            return filterMin.map((product, i) => {
                return <ListOneProduct key={i} product={product} />
            })
        } else {
            let filterMin = this.props.firstProductsPuthname.filter(num => {
                return num.name.toLowerCase().indexOf(this.state.firstHistory) >= 0 ? num : null;
            })
            return filterMin.map((product, i) => {
                return <ListOneProduct key={i} product={product} />
            })
        }
    }
*/