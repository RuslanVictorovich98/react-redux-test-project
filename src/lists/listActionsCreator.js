import {FIND_TO_NAME, EDIT_CATEGORY, CREATE_LIST, GETTING_PRODUCTS, FETCH_USERS_REQUEST, FIRTS_RENDER_PUTHNAME, FIND_TO_NAME_PATHNAME} from './listActions';

export const create = (event) => {
    return {
        type: CREATE_LIST,
        payload: event
    }
}

export const category = (event) => {
    return {
        type: EDIT_CATEGORY,
        payload: event,
    }
}

export const findToName = (event) => {
    return {
        type: FIND_TO_NAME,
        payload: event,
    }
}

export const gettingProducts = (event) => {
    return {
        type: GETTING_PRODUCTS,
        payload: event,
    }
}

export const fetchUsersRequest = () => {
    return {type: FETCH_USERS_REQUEST}
}

export const firstRenderPuthname = (event) => {
    return {
        type: FIRTS_RENDER_PUTHNAME,
        payload: event
    }
}

export const findToNamePathname = (event) => {
    return {
        type: FIND_TO_NAME_PATHNAME,
        payload: event
    }
}