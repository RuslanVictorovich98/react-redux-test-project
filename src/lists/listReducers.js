import {FIND_TO_NAME, EDIT_CATEGORY, CREATE_LIST, GETTING_PRODUCTS, FIRTS_RENDER_PUTHNAME, FIND_TO_NAME_PATHNAME} from './listActions';


const list =  {
    products: [],
    productsFilter: [],
    productsFind: [],
    productsPuthname: [],
    productsPuthnameFind: [],
};

export function mainList (state = list.products, action) { // main
    switch (action.type) {
        case CREATE_LIST:
            list.productsFilter = action.payload;
            return action.payload;

        default:
            return state;
    }
}

export function listReducer (state = list.productsFilter, action) { // list data
    switch (action.type){

        case CREATE_LIST:
            list.productsFilter = action.payload;
            return action.payload;

        case GETTING_PRODUCTS: 
            return action.payload;
            

        case EDIT_CATEGORY:
            return list.productsFilter.filter(num => num.bsr_category.toLowerCase() === action.payload.toLowerCase());

        default:
            return state;
    }
}

export function liftFilter  (state = list.productsFind, action) { // list data after filter
    switch (action.type) {

        case FIND_TO_NAME: 
            return action.payload;

        default:
            return state;
    }
}

export function firstProductsPuthname (state = list.productsPuthname, action) { // data at first load page by category
    switch (action.type) {
        
        case FIRTS_RENDER_PUTHNAME:
            return list.productsFilter.filter(num => num.bsr_category.toLowerCase() === action.payload.toLowerCase());

        default:
            return state;
    }
}

export function firstProductsPuthnameFind (state = list.productsPuthnameFind, action) { // data at first load page by find
    switch(action.type) {

        case FIND_TO_NAME_PATHNAME: 
            return action.payload;

        default: 
            return state;
    }
    
}
