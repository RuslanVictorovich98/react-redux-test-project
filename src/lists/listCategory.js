import React from 'react';
import {connect} from 'react-redux';
import { Router, Link } from "react-router-dom";
import {create, category, gettingProducts} from './listActionsCreator';
import {createBrowserHistory} from 'history';

const history = createBrowserHistory();


class ListCategory extends React.Component {

    renderCategory = () => {

        
        let data =  this.props.main.map((elem) => {
            return elem.bsr_category;
        })
        let categoryData = data.filter((item, pos) => {
            return data.indexOf(item) === pos;
        })
        
        return categoryData.map((elem, i) => {
            return(
                <h4 className="category-list" key={i}  onClick={(e) => {
                    this.props.category(e.currentTarget.innerText);
                }}>
                    <Link to={'/' + elem}>{elem}</Link>
                </h4>);
        })
    }


    getHistory () {
        let findData = this.props.findData;
        if (findData !== '') {
            history.push({search: '?=' + findData});
        } else {
            history.push({search: '?=' + findData});
        }

    }

    render() {
        return(
            <Router history={history}> 
                <h4  className="category-list"><Link to={'/All category'} onClick={ () => this.props.gettingProducts(this.props.main)}>All category</Link></h4> 
                {this.renderCategory()} 
                {/* {this.getHistory()} */}
            </Router>)
        }
    }

const mapStateToProps = (state) => {
    return {
        list: state.list,
        main: state.main,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        create: (i) => dispatch(create(i)),
        category: (i) => dispatch(category(i)),
        gettingProducts: (i) => dispatch(gettingProducts(i)),
    }
}

export default connect (mapStateToProps, mapDispatchToProps) (ListCategory);