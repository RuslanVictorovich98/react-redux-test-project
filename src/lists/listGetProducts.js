import { call, put, takeLatest } from 'redux-saga/effects'// const {put, call} = ReduxSaga.effects;
import {FETCH_USERS_REQUEST} from './listActions';
import { create } from './listActionsCreator';

export function* fetchUser() {
  const url = 'http://localhost:3000/data.json';
  try {
    const data = yield call(() => {
              return fetch(url)
                      .then(res => res.json())
              })
    yield put (create(data.products));
  } catch (error) {
    console.error('fetchUsers return erros: ' + error);
  }        
 }
 
 export default function* GetProduct() {
    yield takeLatest(FETCH_USERS_REQUEST, fetchUser);
  }
