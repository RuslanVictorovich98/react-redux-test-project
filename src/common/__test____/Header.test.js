import React from 'react';
import {shallow} from 'enzyme';
import ConnectedHeader, { Header }  from '../Header';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import { mainList } from '../../ducks/lists';

const initialState = {One: 'one'}
const mockStore = configureStore();
const mockFetchData = jest.fn();
const getHistory = jest.fn();

let store, wrapper, props;

beforeEach(()=>{
    store = mockStore(initialState);
    props = {
        pathname: mockFetchData,
        handleSubmit: getHistory,
        mainList: [{category: ''}],
        search: mockFetchData
    }
    wrapper = shallow(<Header store={store}  {...props} onClickFunction={mockFetchData}/>)
})


    describe('TEST component Header', () => {

        it('TEST spapshot to component', () => {
            // expect(wrapper.find('header').text()).toMatchSnapshot();
            expect(toJson(wrapper)).toMatchSnapshot();
        })

        it('Header component', () => {
            expect(wrapper.exists()).toBe(true) // Returns whether or not any nodes exist in the wrapper.
        })

        it('render initial', () => { 
            expect(wrapper.find('h3')).toHaveLength(1)
            expect(wrapper.find('input')).toHaveLength(1) 
        })

        it('dispatches the `getHistory()` function it receives from props', () => {
            expect(mockFetchData).toHaveBeenCalled() // ожидание с нужным ассертом
        })

        it('should call mockFetchData on input change', () => {
           
            wrapper.find('input').simulate('change', {
                      target: { //1
                        value: ''
                    }});
            expect(mockFetchData).toHaveBeenCalled();
            wrapper.unmount();
          });
    })

    
