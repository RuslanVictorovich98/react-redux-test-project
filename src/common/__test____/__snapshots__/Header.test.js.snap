// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST component Header TEST spapshot to component 1`] = `
<header
  className="App-header"
>
  <Container
    fluid={false}
  >
    <ForwardRef
      noGutters={false}
    >
      <Col
        lg="4"
        sm="4"
      >
        <h3
          className="App-title"
          onClick={[Function]}
        >
          ProductsAPI
        </h3>
      </Col>
      <Col
        lg="6"
        sm="6"
      >
        <input
          className="input-filter form-control"
          defaultValue=""
          id="input-filter"
          onChange={[Function]}
          placeholder="Введите имя...."
          type="search"
        />
      </Col>
    </ForwardRef>
  </Container>
</header>
`;
